### DISCUSSION TITLE
How to create a workbook for both reading and writing data into excel using exceljs?

### DISCUSSION THREAD

**Discussion by santhoshcse** [exceljs/exceljs#2505] [Q&A]
In exceljs, there is option to read data from excel and also to write data.
But there is no option to open a excel file for reading and writing at a time (like r+, w+ mode in c file operation) in the same file.
throws "Error: EBUSY: resource busy or locked, open 'xyz.xlsx'", if i try to do both operation on same file.
**Questions:**
how to close the workbook object (like workbook.close)? 
so that it helps to close the workbook object(xyz.xlsx) which opened for reading and i can open new workbook object(xyz.xlsx) for writing data.
or 
is there any method available for both operation (like : workbook.readWriteExcel())?

**Comment by AniaMakes** [comment]
Hi! Sadly this is not a solution, but yet another person who is wondering about simultaneous read & write functionality. I'm iterating through rows and want the output written in right-most columns of each row.

When I was tackling this in python about 5 years ago, the only solution was to have two copies of the working file; one which is being read and the other one which is being written to. Seems that all these years ahead this approach hasn't changed. 

I think my approach would be to read the spreadsheet first and output the info into an object, and then open the same spreadsheet in a write mode and iterate through the object to insert data, but that does seem rather clunky.

**Comment by Vineeth-Agarwal** [comment]
Solution:
It is very simple as below. Thanks to Sunil Patro for this easiest solution. I am wondering how this solution has not been published or exposed yet. 

https://qavalidation.com/2017/11/readwrite-excel-sheet-using-exceljs.html/

Code snippet from above URL

var Excel = require('exceljs'); 
var wb = new Excel.Workbook(); wb.xlsx.readFile("./test.xlsx").then(function(){     
var sh = wb.getWorksheet("Sheet1");      
sh.getRow(3).getCell(2).value = 32;    
wb.xlsx.writeFile("./test.xlsx");    
console.log("Row-3 \| Cell-2 - "+sh.getRow(3).getCell(2).value);     
console.log(sh.rowCount);    
//Get all the rows data [1st and 2nd column]    
for (i = 1; i <= sh.rowCount; i++) {         console.log(sh.getRow(i).getCell(1).value);        
console.log(sh.getRow(i).getCell(2).value);    
}})

**Comment by xitanggg** [comment]
I can read and write to the same file okay by doing the following:
```javascript
const ExcelJS = require('exceljs');
const workbook = new ExcelJS.Workbook();
const fileName = './myFile.xlsx';
(async function () {
	// Read file and Sheet 1
	const worksheet = (await workbook.xlsx.readFile(fileName)).getWorksheet(
		'Sheet1'
	);
	// Change row 1 cell 1 to 'newValue'
	worksheet.getRow(1).getCell(1).value = 'newValue';
	// Write to same file
	try {
		await workbook.xlsx.writeFile(fileName);
	} catch (error) {
		console.log('Write file fails: ', error);
	}
})();
```

If you open the file with excel while performing `writeFile`, it will throw the error `Resource busy`. I believe excel locks the file's write access once you open the file with it. I add a warning at the beginning of my code to remind me to close the file.
```javascript
const readline = require('readline-sync');
const msg = `Warning: Make sure to CLOSE excel file before proceeding!`;
readline.question(msg);
```
